control-plane/agents/src/bin/ha/cluster/switchover.rs

```diff
@@ -257,59 +294,77 @@ impl SwitchOverEngine {
-                drop(lo);
-                let retry_delay_min = std::time::Duration::from_secs(1);
-                let retry_delay_max = std::time::Duration::from_secs(384);
-                let mut retry_delay = retry_delay_min;
                 loop {
                     // TODO: handle error handling properly
                     let result = match q.stage {
                         Stage::Init => q.initialize(&self.etcd).await,
                         Stage::RepublishVolume => q.republish_volume(&self.etcd).await,
                         Stage::PublishPath => q.publish_path(&self.etcd, &self.nodes).await,
                         Stage::DeleteTarget => q.delete_target(&self.etcd).await,
+                        Stage::Errored => match q.errored_switchover(&self.etcd).await {
+                            Ok(_) => break,
+                            Err(e) => Err(e),
+                        },
                         Stage::Successful => match q.delete_switchover(&self.etcd).await {
                             Ok(_) => break,
                             Err(e) => Err(e),
                         },
-                        _ => break,
                     };
                     match result {
                         Ok(_) => {
-                            // reset retry delay back to the start.
-                            retry_delay = retry_delay_min;
+                            // reset retry count back to the start after successfully completing a
+                            // stage.
+                            q.retry_count = 0;
                         }
                         Err(_) => {
-                            retry_delay = (retry_delay * 3).min(retry_delay_max);
-                            tokio::time::sleep(retry_delay).await;
+                            q.retry_count += 1;
+                            let retry_delay = if q.retry_count < NUM_OF_FAST_REQUEUE {
+                                FAST_REQUEUE
+                            } else {
+                                SLOW_REQUEUE
+                            };
+                            let _ =
+                                self.initiate(q.clone(), Some(Duration::from_secs(retry_delay)));
+                            info!(volume.uuid=%q.volume_id, "Sending failed Switchover request back to the channel");
+                            break;
                         }
                     }
                 }
             }
         }
     }
 
-    pub fn initiate(&self, req: SwitchOverRequest) -> anyhow::Result<()> {
-        self.channel
-            .send(req)
-            .map_err(|error| anyhow::anyhow!("Failed to send switchover request: {error}"))
+    /// Sends Switchover request to the Channel after sleeping for a specified duration if
+    /// specified.
+    pub fn initiate(
+        &self,
+        req: SwitchOverRequest,
+        duration: Option<Duration>,
+    ) -> anyhow::Result<()> {
+        let clone_self = self.clone();
```

Comment:
Could we clone just the channel since we don't seem to use self after?

Reply:
Incorporated